### DISCUSSION TITLE
Not able to achieve authentication in next.js app router.

### DISCUSSION THREAD

**Discussion by heyarviind** [pocketbase/pocketbase#4430] [Q&A]
I am trying to achieve authentication in next.js app router but for some reason, it is not working at all. My flow is, the user submits the form on frontend and email and password is sent to the backend where pocketbase authenticates the user and set the cookie in the browser

```
const authData = await pocketBase
      .collection("users")
      .authWithPassword(email, password);

    // set cookie on the client
    cookies().set({
      name: "pb_auth",
      value: authData.token,
      httpOnly: false,
      secure: false,
      // expires in 6 months
      maxAge: 60 * 60 * 24 * 180,
      path: "/",
      sameSite: "lax",
    });
```

I have two pocketBase instances, one for the server and one for the browser as shown below:

*For Server*
```
import { cookies } from "next/headers";
import PocketBase from "pocketbase";

export async function createServerClient() {
  if (!process.env.NEXT_PUBLIC_PB_ENDPOINT) {
    throw new Error("Pocketbase API url not defined !");
  }

  if (typeof window !== "undefined") {
    throw new Error(
      "This method is only supposed to call from the Server environment"
    );
  }

  const client = new PocketBase(process.env.NEXT_PUBLIC_PB_ENDPOINT);

  const pb_auth = cookies().get("pb_auth");

  if (pb_auth) {
    const authCookie = cookies().get("pb_auth");

    if (authCookie) {
      client.authStore.loadFromCookie(`pb_auth=${authCookie.value}`, "pb_auth");

      console.log(
        client.authStore,
        `${authCookie.name}=${authCookie.value}`,
        client.authStore.isValid
      );

      try {
        // get an up-to-date auth store state by veryfing and refreshing the loaded auth model (if any)
        client.authStore.isValid &&
          (await client.collection("users").authRefresh());
      } catch (_) {
        // clear the auth store on failed refresh
        client.authStore.clear();
      }
    }
  } else {
    client.authStore.clear();
    console.log("No auth cookie found!");
  }

  return client;
}

```

For Browser:
```
import PocketBase from "pocketbase";

let singletonClient: any;

export function createBrowserClient() {
  if (!process.env.NEXT_PUBLIC_PB_ENDPOINT) {
    throw new Error("Pocketbase API url not defined !");
  }

  const createNewClient = () => {
    return new PocketBase(process.env.NEXT_PUBLIC_PB_ENDPOINT);
  };

  const _singletonClient = singletonClient ?? createNewClient();

  if (typeof window === "undefined") return _singletonClient;

  if (!singletonClient) singletonClient = _singletonClient;

  return singletonClient;
}
```

At this time, whenever I do `client.authStore.loadFromCookie(`pb_auth=${authCookie.value}`, "pb_auth");` and then check `client.authStore.isValid` it returns false.

Am I missing something?

**Comment by ganigeorgiev** [comment]
You could check some of the user provided examples in https://github.com/pocketbase/js-sdk/issues/69.


In the above code the `if (pb_auth) {...` check seems unnecessary because you are doing the same with `if (authCookie) {...`. Additionally, if the cookie value is just the token, you don't have to construct a cookie string and call `authStore.loadFromCookie(...)`, instead you can call directly `authStore.save(token)`.

Also the above code doesn't show, or at least it is not clear, when `cookies().set({` is invoked and how (_you've mentioned that you are handling the login on the server-side but at the same time you set the cookie from the browser?_).
The general idea is that the cookie must be updated every time when the store changes, otherwise your `authRefresh()`, `authStore.clear()`, etc. doesn't seem that they will have any effect.

With that said, I don't recommend using JavaScript SSR with PocketBase, especially with Next.js and with mixed browser/server access maintaining regular users auth as it seems in your case. I've tried to explain in more details why in https://github.com/pocketbase/pocketbase/discussions/4412#discussioncomment-8575834, https://github.com/pocketbase/pocketbase/discussions/3792#discussioncomment-7621130.
